unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,432 @@
+use cid::Cid;
+
+use crate::pb::{FlatUnixFs, PBLink, UnixFs, UnixFsType};
+use quick_protobuf::{MessageWrite, Writer};
+use std::borrow::Cow;
+use std::fmt;
+use std::num::NonZeroUsize;
+
+use sha2::{Digest, Sha256};
+
+/// File tree builder
+pub struct FileAdder {
+    chunker: Chunker,
+    block_buffer: Vec<u8>,
+    // the index in the outer vec is the height (0 == leaf, 1 == first links to leafs, 2 == links
+    // to first link blocks)
+    unflushed_links: Vec<Vec<(Cid, u64, u64)>>,
+}
+
+impl std::default::Default for FileAdder {
+    fn default() -> Self {
+        Self::with_chunker(Chunker::Size(1024 * 256))
+    }
+}
+
+impl fmt::Debug for FileAdder {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "FileAdder {{ chunker: {:?}, block_buffer: {}/{}, unflushed_links: {} }}",
+            self.chunker,
+            self.block_buffer.len(),
+            self.block_buffer.capacity(),
+            self.unflushed_links
+                .iter()
+                .map(|l| l.len().to_string())
+                .collect::<Vec<_>>()
+                .join("/")
+        )
+    }
+}
+
+impl FileAdder {
+    /// Creates `FileAdder` with the given chunker. Typically one could just call
+    /// `FileAdder::default()`.
+    pub fn with_chunker(chunker: Chunker) -> Self {
+        let hint = chunker.size_hint();
+        FileAdder {
+            chunker,
+            block_buffer: Vec::with_capacity(hint),
+            unflushed_links: Default::default(),
+        }
+    }
+
+    /// Returns a likely amount of buffering the file adding works the best.
+    pub fn size_hint(&self) -> usize {
+        self.chunker.size_hint()
+    }
+
+    /// Called to push new file bytes into the tree builder.
+    ///
+    /// Returns the newly created blocks (at most 2) and their respective Cids, and the amount of
+    /// `input` consumed.
+    pub fn push(
+        &mut self,
+        input: &[u8],
+    ) -> Result<(impl Iterator<Item = (Cid, Vec<u8>)>, usize), ()> {
+        // case 0: full chunk is not ready => empty iterator, full read
+        // case 1: full chunk becomes ready, maybe short read => at least one block
+        //     1a: not enough links => iterator of one
+        //     1b: link block is ready => iterator of two blocks
+
+        let (accepted, ready) = self.chunker.accept(input, &self.block_buffer);
+        self.block_buffer.extend_from_slice(accepted);
+        let written = accepted.len();
+
```

Comment:
Minor 10MB/s increase if a "less slow path" is added here where `self.block_buffer.is_empty() && ready`, as in the caller uses a `buffer.len() >= self.size_hint()`. BufReader doesn't do this normally, with `256 * 1024` buffer it's `fill_buf` will only issue a single read and the result will be probably the platform stdin buffer size or anything below it (I got 4096 multiples 1 and 32).

Reply:
Doing full `size_hint` reads with two `Vec`s with two threads produces similar speed as the single threaded version (passing the buffers and back forth with two channels). Probably the bottleneck is the many allocations while writing..

Reply:
After a few flamegraphs later: for the raw ingestion rate doesn't really matter on how clever the stdin reading is. It depends on the sha2 (quite surprisingly).

I tried these stdin reading approaches:
 * std::io::BufReader (which issues a single `read`)
 * full size_hint sized vecs (many `read`s)
 * another thread with std::sync::mpsc::channel for communication with n size_hint vecs (`N >= 2`)

I wont complicate this PR with the above but I will push that as somewhere to bitrot and update this later with a link.

All of the stdin reading approaches (with this "less-slow-path" fix) produced about ~230MBps ingestion rate for a 5GB test input. On my laptop I also have firefox running with gazillion tabs and so on.

Upgrading sha2 from 0.8.1 to 0.9.1 for this experiment yields ~1GBps ingestion rate.

This leads me to believe that this copying heavy but simple code is good enough (tm).